src/meteors/attr/noise_tunnel.py

```diff
@@ -1,68 +1,583 @@
+            baseline (torch.Tensor): A baseline tensor to replace the perturbed bands.
+            n_samples (int): A number of samples to be drawn - number of perturbed inputs to be generated.
+            perturbation_prob (float, optional): A probability that each band will be perturbed intependently. Defaults to 0.5.
+            num_perturbed_bands (int | None, optional): A number of perturbed bands in the whole image. If set to None, the bands are perturbed with probability `perturbation_prob` each. Defaults to None.
+
+        Returns:
+            torch.Tensor: A perturbed tensor, which contains `n_samples` perturbed inputs.
+        """
+        # validate the baseline against the input
+        if baseline.shape != input.shape:
+            raise ValueError(f"Baseline shape {baseline.shape} does not match input shape {input.shape}")
+
+        if n_samples < 1:
+            raise ValueError("Number of perturbated samples to be generated must be greater than 0")
+
+        if perturbation_prob < 0 or perturbation_prob > 1:
+            raise ValueError("Perturbation probability must be in the range [0, 1]")
+
+        # the perturbation
+        perturbed_input = input.clone().unsqueeze(0)
+        # repeat the perturbed_input on the first dimension n_samples times
+        perturbed_input = perturbed_input.repeat_interleave(n_samples, dim=0)
+
+        n_samples_x_channels_shape = (
+            n_samples,
+            input.shape[0],
+        )  # shape of the tensor containing the perturbed channels for each sample
+
+        channels_to_be_perturbed: torch.Tensor = torch.zeros(n_samples_x_channels_shape, device=input.device).bool()
+
+        if num_perturbed_bands is None:
+            channel_perturbation_probabilities = (
+                torch.ones(n_samples_x_channels_shape, device=input.device) * perturbation_prob
+            )
+            channels_to_be_perturbed = torch.bernoulli(channel_perturbation_probabilities).bool()
+
+        else:
+            if num_perturbed_bands < 0 or num_perturbed_bands > input.shape[0]:
+                raise ValueError(
+                    f"Cannot perturb {num_perturbed_bands} bands in the input with {input.shape[0]} channels. The number of perturbed bands must be in the range [0, {input.shape[0]}]"
+                )
+
+            channels_to_be_perturbed = torch_random_choice(
+                input.shape[0], num_perturbed_bands, n_samples, device=input.device
+            )
+
+        # now having chosen the perturbed channels, we can replace them with the baseline
+
+        reshaped_baseline = baseline.unsqueeze(0).repeat_interleave(n_samples, dim=0)
+        perturbed_input[channels_to_be_perturbed] = reshaped_baseline[channels_to_be_perturbed]
+
+        perturbed_input.requires_grad_(True)
+
+        return perturbed_input
+
+    def attribute(
+        self,
+        inputs: list[HSI] | HSI,
```

Comment:
base class should take torch.Tensor as input

Reply:
this is a bigger problem connected to the structure of BaseNoiseTunnel - we need to decide whether this class should handle HSI objects or work with raw tensors, similarily to the captum ones